### ISSUE TITLE
XLR v9.8.0 - Launch Job Template from this plugin raises ImportError: No module named ordered_dict in <script> at line number 67

### ISSUE THREAD

**Issue by rubenryhan** [xebialabs-community/xlr-ansible-tower-plugin#8]
Hi Team,
After upgrading to XLR v9.8.0 - "Launch Job Template"  (same issue for launch Job/synchronize inventory as well) from this plugin raises ImportError: No module named ordered_dict in <script> at line number 67. 
Appreciate your feedback and time.

**Comment by rubenryhan** [comment]
Hi,
Any input on how to work around/resolve this issue? It seems to be happening on 9.7.6 version of XLR as well? on 9.7.6 the error is as follows - javax.script.ScriptException: ImportError: No module named ordered_dict .

**Comment by rubenryhan** [comment]
Here is the full error stack - 
c.x.x.s.jython.JythonScriptService - Exception during execution
javax.script.ScriptException: ImportError: No module named ordered_dict in <script> at line number 67
        at org.python.jsr223.PyScriptEngine.scriptException(PyScriptEngine.java:222)
        at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:59)
        at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:31)
        at com.xebialabs.xlrelease.script.Jsr223ScriptExecutor.lambda$doPrivileged$0(Jsr223ScriptExecutor.java:72)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.xebialabs.xlrelease.script.Jsr223ScriptExecutor.doPrivileged(Jsr223ScriptExecutor.java:72)
        at com.xebialabs.xlrelease.script.jython.JythonScriptExecutor.evalScriptPrivileged(JythonScriptExecutor.java:67)
        at com.xebialabs.xlrelease.script.jython.JythonScriptExecutor.evalScriptWithUtilities(JythonScriptExecutor.java:80)
        at com.xebialabs.xlrelease.script.jython.JythonScriptExecutor.evalScriptWithApi(JythonScriptExecutor.java:56)
        at com.xebialabs.xlrelease.script.DefaultScriptService.executeScript(DefaultScriptService.java:857)
        at com.xebialabs.xlrelease.script.DefaultScriptService.doExecuteScript(DefaultScriptService.java:569)
        at com.xebialabs.xlrelease.script.DefaultScriptService.lambda$executeCustomScriptTask$1(DefaultScriptService.java:225)
        at com.xebialabs.deployit.plumbing.scheduler.Scheduler.lambda$withLoggedExceptions$0(Scheduler.java:69)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.python.core.PyException: ImportError: No module named ordered_dict
        at org.python.core.PyException.doRaise(PyException.java:211)
        at org.python.core.Py.makeException(Py.java:1638)
        at org.python.core.Py.makeException(Py.java:1642)
        at org.python.core.Py.makeException(Py.java:1646)
        at org.python.core.Py.makeException(Py.java:1650)
        at org.python.pycode._pyx21.process$2(<script>:63)
        at org.python.pycode._pyx21.call_function(<script>)
        at org.python.core.PyTableCode.call(PyTableCode.java:173)
        at org.python.core.PyBaseCode.call(PyBaseCode.java:134)
        at org.python.core.PyFunction.__call__(PyFunction.java:416)
        at org.python.pycode._pyx21.f$0(<script>:67)
        at org.python.pycode._pyx21.call_function(<script>)
        at org.python.core.PyTableCode.call(PyTableCode.java:173)
        at org.python.core.PyCode.call(PyCode.java:18)
        at org.python.core.Py.runCode(Py.java:1687)
        at org.python.core.__builtin__.eval(__builtin__.java:497)
        at org.python.core.__builtin__.eval(__builtin__.java:501)
        at org.python.util.PythonInterpreter.eval(PythonInterpreter.java:255)
        at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:57)
        ... 18 common frames omitted

**Comment by rubenryhan** [comment]
An update on this issue, after further investigation, it appears this error can simply be reproduced, if we try to “import requests” in a jython script task -
Exception during execution:
javax.script.ScriptException: ImportError: No module named ordered_dict in

any ideas how this can be worked around?

**Comment by bmoussaud** [comment]
@rubenryhan can you try install it (not tested) 
1. Download https://files.pythonhosted.org/packages/53/25/ef88e8e45db141faa9598fbf7ad0062df8f50f881a36ed6a0073e1572126/ordereddict-1.1.tar.gz
2. extract /ordereddict-1.1/ordereddict.py in the ext/ folder. (Maybe you need to restart the server)

**Comment by rubenryhan** [comment]
@bmoussaud  still getting same error after copying ordereddict.py to XLR_HOME/ext folder and restarting XLR server. Looking forward to your response.
regards

**Comment by rubenryhan** [comment]
Hi team,
Appreciate your time, I have worked out what the issue was in this case, it is incompatibility between urllib3 and requests module. I have not done extensive testing yet, but here is what I have so far, in my environment I have both xlr-ansible-tower-plugin and [xlr-xlrelease-plugin](https://github.com/xebialabs-community/xlr-xlrelease-plugin) installed.

`1.15.1`  <<< version of urllib3 in XLR v9.7.2 (working environment with both plugins installed)
`1.25.9`  <<< version of urllib3 in XLR v9.8.0/9.7.6 (issue occurs in this environment with both plugins installed)

`2.19.1`  <<< version of requests module shipped with xlr-xlrelease-plugin, according to documentations that are around urllib3 is only supported up to v1.23.x for requests module 2.19.1 

In my case, I have replaced the requests module within xlr-xlrelease-plugin-2.3.6.jar file with requests module (v 2.24.0) , and that has resolved the issue across the board with ImportError: No module named ordered_dict. Further to that, if I do not install xlr-xlrelease-plugin , then I do not have this issue at all anyway and ansible-tower plugin also works fine in that case. Also, I have tested removing requests module from xlr-xlrelease-plugin-2.3.7.jar file and that also resolves the issue.

regards,
